### PR TITLE
Support coercing `utf8` to `interval` and `timestamp` (including arguments to `date_bin`)

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it~
- [x] builds on https://github.com/apache/arrow-datafusion/pull/5115

# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/4853

# Rationale for this change

Rationale:

`date_bin` with a string value does not work and gives a hard to understand message:

```sql
❯  select date_bin('1 hour', column1, TIMESTAMP '2001-01-01 00:00:00Z')
from (values
  (timestamp '2022-01-01 00:00:00'),
  (timestamp '2022-01-01 01:00:00'),
(timestamp '2022-01-02 00:00:00')
) as sq;
Plan("Coercion from [Utf8, Timestamp(Nanosecond, None), Timestamp(Nanosecond, None)] to the signature Exact([Interval(DayTime), Timestamp(Nanosecond, None), Timestamp(Nanosecond, None)]) failed.")
```

# What changes are included in this PR?
1. Add coercion so that `Utf8` constants can be automatically coerced into Intervals (same as today)
2. Add coercion so that `Utf8` can be automatically coerced into Timestamps
3. Tests

# Are these changes tested?
Yes

# Are there any user-facing changes?

Yes, strings are now automatically coerced into `Timetamp`s and `Intervals`

Note this behavior is consistent with Postgres:

```sql
postgres=# SELECT DATE_BIN(INTERVAL '15 minutes', TIMESTAMP '2022-08-03 14:38:50Z', TIMESTAMP '1970-01-01T00:00:00Z') AS res;
         res
---------------------
 2022-08-03 14:30:00
(1 row)

postgres=# SELECT DATE_BIN('15 minutes', TIMESTAMP '2022-08-03 14:38:50Z', TIMESTAMP '1970-01-01T00:00:00Z') AS res;
         res
---------------------
 2022-08-03 14:30:00
(1 row)

postgres=# SELECT DATE_BIN('15 minutes', '2022-08-03 14:38:50Z', '1970-01-01T00:00:00Z') AS res;
          res
------------------------
 2022-08-03 14:30:00+00
(1 row)
```

### CODE REVIEW

File: datafusion/optimizer/src/type_coercion.rs [hunk 1]
```diff
@@ -553,10 +553,22 @@ fn coerce_arguments_for_signature(
...
-        .map(|(i, expr)| expr.clone().cast_to(&new_types[i], schema))
+        .map(|(i, expr)| cast_expr(expr, &new_types[i], schema))
         .collect::<Result<Vec<_>>>()
 }
 
+/// Cast `expr` to the specified type, if possible
+fn cast_expr(expr: &Expr, to_type: &DataType, schema: &DFSchema) -> Result<Expr> {
+    // Special case until Interval coercion is handled in arrow-rs
+    // https://github.com/apache/arrow-rs/issues/3643
+    match (expr, to_type) {
```
Comment (alamb) [inline datafusion/optimizer/src/type_coercion.rs:564]: This only supports literals at the moment -- tracked https://github.com/apache/arrow-rs/issues/3643 for the more general case of coercing columns

**Comment by alamb** [comment]
cc @waitingkuo  and @comphead

File: datafusion/core/tests/sqllogictests/test_files/timestamps.slt [hunk 2]
```diff
@@ -76,6 +76,18 @@ SELECT DATE_BIN(INTERVAL '15 minutes', TIMESTAMP '2022-08-03 14:38:50Z', TIMESTA
 ----
 2022-08-03T14:30:00
 
+# Can coerce string interval arguments
+query T
+SELECT DATE_BIN('15 minutes', TIMESTAMP '2022-08-03 14:38:50Z', TIMESTAMP '1970-01-01T00:00:00Z') AS res
```
Comment (comphead) [inline datafusion/core/tests/sqllogictests/test_files/timestamps.slt:81]: small nit: if AS res is needed?

**Comment by comphead** [comment]
LGTM, not permitted to approve the PR :)

File: datafusion/core/tests/sqllogictests/test_files/timestamps.slt (hunk 2 above)
Comment (alamb) [reply datafusion/core/tests/sqllogictests/test_files/timestamps.slt:81]: I think you are right -- `res` is not needed. I will remove them

Comment (alamb) [reply datafusion/core/tests/sqllogictests/test_files/timestamps.slt]: In 7a7fd4b92

**Pull request merged**